External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <atomic>
+#include <FEXCore/Utils/Allocator.h>
+
+#include <chrono>
+#include <cstddef>
+#include <list>
+#include <mutex>
+
+namespace FEXCore::Utils {
+  class ThreadPoolAllocator final {
+    public:
+      template<typename T>
+      struct AllocationInfo {
+        T Ptr;
+        size_t Size;
+      };
+
+      using ClockType = std::chrono::high_resolution_clock;
+
+      struct MemoryBuffer {
+        void* Ptr;
+        size_t Size;
+        std::chrono::time_point<ClockType> LastUsed;
+        std::atomic<bool> ClientClaimed;
+        std::atomic<bool> *CurrentClientOwnedFlag{};
+      };
+
+      using ContainerType = std::list<MemoryBuffer*>;
+
+      ContainerType::iterator ClaimBuffer(size_t Size, std::atomic<bool> *CurrentClientFlag) {
+        auto Buffer = ClaimBufferImpl(Size);
+        (*Buffer)->ClientClaimed = true;
+        (*Buffer)->CurrentClientOwnedFlag = CurrentClientFlag;
+        CurrentClientFlag->store(true);
+        return Buffer;
+      }
+
+      void UnclaimBuffer(ContainerType::iterator Buffer) {
+        std::unique_lock lk {AllocationMutex};
+        UnclaimBufferImpl(Buffer);
+      }
+
+      void DisownBuffer(ContainerType::iterator Buffer) {
+        // Client still owns the buffer but isn't using it
+        // Allows us to claim it back if necessary
+        (*Buffer)->LastUsed = ClockType::now();
+        (*Buffer)->ClientClaimed.store(false);
+      }
+
+      ~ThreadPoolAllocator() {
+        // Cleanup any buffers that are remaining
+        for (auto it = Buffers.begin(); it != Buffers.end(); ++it) {
```

Comment:
range-based for loops can be used here and in a couple of other places.

Reply:
Range based used for the ones that can. The other ones still need the iterators to be passed around.